subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/CleanupActionFactory.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.cache.internal;
+
+import org.gradle.cache.CleanupAction;
+import org.gradle.cache.PersistentCache;
+import org.gradle.internal.operations.BuildOperationContext;
+import org.gradle.internal.operations.BuildOperationExecutor;
+import org.gradle.internal.operations.RunnableBuildOperation;
+import org.gradle.internal.progress.BuildOperationDescriptor;
+import org.gradle.internal.reflect.Instantiator;
+
+public class CleanupActionFactory {
+    private final Instantiator instantiator;
+    private final BuildOperationExecutor buildOperationExecutor;
+
+    public CleanupActionFactory(Instantiator instantiator, BuildOperationExecutor buildOperationExecutor) {
+        this.instantiator = instantiator;
+        this.buildOperationExecutor = buildOperationExecutor;
+    }
+
+    public CleanupAction create(Class<? extends CleanupAction> cleanupType, Object... args) {
```

Comment:
wouldn't we just pass the `CleanupAction` to be decorated in? do we get anything from instantiating it here?